### PR TITLE
Lab 03 - Interrupts and timers

### PR DESCRIPTION
**Pull request by irina-b-dev**
### Pull Request Overview

This pull request adds  lab03 - Interrupts and timers

### TODO or Help Wanted

This pull request needs  review

### Build

- [x] Ran `npm build`.

### Author

Signed-off-by: Irina Bradu - [irina.bradu@wyliodrin.com](mailto:irina.bradu@wyliodrin.com)

### CODE REVIEW

File: docs/lab/03/03.md [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+```rust
+
+extern "C" fn handle_SysTick_Exception(){
+
+    //do something after number_of_ticks were counted
+  
+}
+
+//don't forget to link this function at the appropriate address in the vector table
+
```
[the remark this answers is not in the document]
Comment (alexandruradovici) [reply docs/lab/03/03.md:477]: Register the function first, after show the examples of enabling SysTick.

File: docs/lab/03/03.md [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+### Await the Timer 
+Use the await keyword on the timer's delay function to pause the execution of the current task until the timer expires. This is a non-blocking wait, so other async tasks can run during this time.
+```rust
+Timer::after(Duration::from_millis(100)).await;
+//wait in current task 100 milliseconds
+```
+
+Once the timer expires, the task that awaited the timer resumes execution from the point it awaited.
+
+## Inspect binaries
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:536]: I would add this section after explaining the interrupt vector.

File: docs/lab/03/03.md [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+
+```bash
+cargo objdump --target thumbv6m-none-eabi -- -disassemble -j .section_name
+```
+
+:::info
+
+cargo objdumb and rust-objdump tools use llvm's objdump tool.
+Some terminals do not parse parameters proxied to llvm-objdump command.
+If this happens please use llvm-objdump script directly like so
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:594]:
```suggestion
`cargo objdump` and `rust-objdump` tools use llvm's `objdump` tool.
Some terminals do not parse parameters proxied to `llvm-objdump` command.
If this happens please use `llvm-objdump` script directly like so
```

File: docs/lab/03/03.md [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+
+
+## Resources
+
+1. **Raspberry Pi Ltd**, *[RP2040 Datasheet](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf)*
+2. [Cortex-M0+ Devices Generic User Guide](https://developer.arm.com/documentation/dui0662/latest/)
+3. [Embassy documentation](https://embassy.dev/book/dev/runtime.html)
+
+
+## What is an exception ?
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:19]:
```suggestion
## What is an exception? 
```

File: docs/lab/03/03.md [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+|-----------|----------|--------------|
+| Reset | -15 | Triggered by the system at startup to start the software |
+| NMI | -14 | Non Maskable Interrupts, an external interrupt that cannot be ignored, usually used for low latency peripheral needs the attention of the MCU |
+| HardFault | -13 | Triggered by the MCU in case of a fault (div by 0, memory fault, ...) |
+| SVC | -5 | Supervisor call, triggered usually by a process running on top of the embedded operating system when it wants to make a system call |
+| PendSV | -2 | Used for pending system calls |
+| SysTick | -1 | Triggered by a periodic timer, usually used by an embedded operating system for context switch |
+| Interrupt (IRQ) | 0 and above| Triggered by a peripheral, or generated by a software request
+
+## What is an interrupt ?
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:35]:
```suggestion
## What is an interrupt?
```

File: docs/lab/03/03.md [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+
+
+### Interrupts on RP2040
+
+The RP2040 chip has two cores (processors), and each core has its own NVIC. Each core's NVIC is connected to the same set of hardware interrupt lines with one exception : IO Interrupts. In the RP2040, IO interrupts are organized by banks, and each core has its own set of IO interrupts for each bank. The IO interrupts for each core are completely independent. For instance, Processor 0 (Core 0) can be interrupted by an event on GPIO pin 0 in bank 0, while Processor 1 (Core 1) can be interrupted by a separate event on GPIO pin 1 in the same bank. Each processor responds only to its own interrupts, allowing them to operate independently or to handle different tasks simultaneously without interfering with each other.
+
+![rp2040 subsystem](images/rpsubsystem.png)
+
+#### Supported interrupt request signals (IRQ)
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:62]:
```suggestion
#### Available interrupt request signals (IRQ)
```

File: docs/lab/03/03.md [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+
+On RP2040, only the lower 26 IRQ signals are connected on the NVIC, as seen in the table below, and IRQs 26 to 31 are tied to zero (never firing).
+The core can still be forced to enter the relevant interrupt handler by writing bits 26 to 31 in the NVIC ISPR register.
+
+![Supported IRQ](images/IRQrp2040.png)
+
+The priority order is determined for these signals is determined by :
+
+- First, the dynamic priority level configured per interrupt by the NVIC_IPR0-7 registers. The Cortex-M0+ implements
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:71]:
```suggestion
- First, the dynamic priority level configured per interrupt by the `NVIC_IPR0-7` registers. The Cortex-M0+ implements
```

File: docs/lab/03/03.md [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+- Falling Edge: Interrupt occurs when a pin transitions from HIGH to LOW.
+
+![alt text](images/interrupt_ll_edge_high.png)
+
+## How to setup an interrupt (or exception) -- bare metal approach
+
+The following steps won't be enough for full bare metal setup of the interrupts. It is only a model that's meant to help understanding this process which is abstracted in higher level frameworks like embassy.
+
+#### 1. Write an interrupt handler
+An interrupt handler is a function that should be executed when a the interrupt signal is triggered. This function must have the signature **fn()** without any arguments or return value.
```
Comment (alexandruradovici) [inline docs/lab/03/03.md:93]:
```suggestion
An interrupt handler is a function that should be executed when a the interrupt signal is triggered. This function must have the signature `fn()` without any arguments or return value.
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: docs/lab/03/03.md [hunk 9]
```diff
@@ -0,0 +1,626 @@
...
+need to return the processor to what it was doing before before the interrupt request was called. This part is called the stack frame and it includes registers 0 through 3, register 12, the link register, the program counter and the program status register.
+
+![alt text](images/stack_frame.png)
+
+
+ARM Cortex-M microcontrollers use a similar concept to the IVT called Nested Vectored Interrupt Controller (NVIC).The NVIC is specifically designed to handle these interrupts more efficiently. Interrupt addresses in the NVIC memory region are set according to their priority: the lower the address, the higher the priority. As suggested by the "Nested" in its name, the NVIC supports nested interrupts. This means that if a higher priority interrupt occurs while another interrupt is being processed, the controller can pause the current interrupt service routine (ISR), handle the higher priority interrupt, and then resume the interrupted ISR. This feature is crucial for responsive and real-time processing.
```
[the remark this answers is not in the document]
Comment (irina-b-dev) [reply docs/lab/03/03.md]: removed comparison

**Review by alexandruradovici — APPROVED** [review]
The code for `embedded-hal` still needs some modifications, we'll have another pull request,

**Review by irina-nita — APPROVED** [review]

**Pull request merged**